lib/thread.c

```diff
@@ -41,6 +41,14 @@ DEFINE_MTYPE_STATIC(LIB, THREAD_STATS,  "Thread stats")
 #include <mach/mach_time.h>
 #endif
 
+static unsigned char wakebyte = 0x01;
+
+#define ACQUIRE_MASTER(m) \
+  do { \
+    while (pthread_mutex_trylock (&m->mtx)) \
+      write (m->io_pipe[1], &wakebyte, 1); \
+  } while (0);
+
```

Comment:
> > This will block in write() after filling up the kernel's pipe buffer, i.e 4096 useless syscalls later.

> No. The pipe should be set to nonblocking. My understanding is that when the buffer fills up it'll just return EAGAIN or similar and continue to spinlock. Correct me if I'm wrong. Clearly I missed one fcntl() call on the write fd; thank you for catching that.

If you make it nonblocking it will call write() for 1829410929 times until the kernel decides to schedule the other thread and actually make progress on the select(). This will be especially bad on uniprocessor systems because the select isn't even running in parallel, so it'll take up the entirety of a scheduler slice before the select() even has a chance to proceed. And note that "waiting in poll" is a state a thread will frequently be in, so this produces pretty large negative impact.

It would in fact be better to change the code to write 4096 bytes and keep it blocking. That way there's 2 write calls at worst, the second of which will immediately block. It'll still be less overhead than writing 4096 * 1 byte.

> > The better design is for thread_fetch() to not hold m->mtx during the fd_select call.

> I went this direction initially.

Why the change? It works...
https://github.com/opensourcerouting/frr/blob/threads/lib/thread.c#L1337

> > There is no reason for it to do so; none of the working data of the fd_select call can be modified by another thread.

> This is incorrect and is precisely the problem I ran into. The thread master's fdset / pollfds are passed directly -- not copied -- into poll or select.

Copy it under lock, then release the lock. The select() call does already copy it (because fd_set is a value type), it's just the poll() path that has the issue.